### PR TITLE
Split transitive dev_dependencies from other transitive dependencies

### PR DESCRIPTION
**Pull request by sigurdm**
Fixes https://github.com/dart-lang/pub/issues/2417

### CODE REVIEW

File: test/outdated/goldens/newer_versions.txt [hunk 1]
```diff
@@ -126,6 +134,8 @@ foo           *1.2.3   *1.3.0      3.0.0       3.0.0
 transitive dependencies
 transitive    *1.2.3   2.0.0       2.0.0       2.0.0   
 
+transitive dev_dependencies
+
```
Comment (jonasfj) [inline test/outdated/goldens/newer_versions.txt]: empty section, but it doesn't say `all up to date`

Comment (jonasfj) [reply test/outdated/goldens/newer_versions.txt]: Also if we have  `--no-dev-dependencies` let's not show this section.

Comment (sigurdm) [reply test/outdated/goldens/newer_versions.txt]: Hmm, both good points!

File: lib/src/command/outdated.dart [hunk 2]
```diff
@@ -181,6 +194,25 @@ class OutdatedCommand extends PubCommand {
...
+      Package root, Iterable<PackageId> resolution) async {
+    final mapping =
+        Map<String, PackageId>.fromIterable(resolution, key: (id) => id.name);
+    final visited = <String>{root.name};
+    final toVisit = [...root.dependencies.keys];
+    while (toVisit.isNotEmpty) {
+      final name = toVisit.removeLast();
+      if (!visited.add(name)) continue;
+      final id = mapping[name];
+      if (id == null) print('id $name mapping $mapping');
```
Comment (jonasfj) [inline lib/src/command/outdated.dart]: print?

Comment (jonasfj) [reply lib/src/command/outdated.dart]: Throw an `AssertionError`

Comment (sigurdm) [reply lib/src/command/outdated.dart]: Arggg! I thought I deleted that

Comment (sigurdm) [reply lib/src/command/outdated.dart]: Remove

File: test/outdated/goldens/newer_versions.txt (hunk 1 above)
Comment (sigurdm) [reply test/outdated/goldens/newer_versions.txt]: Done, Done

**Pull request merged**